go/ql/lib/semmle/go/security/IncorrectIntegerConversionLib.qll

```diff
@@ -150,49 +151,169 @@ deprecated class ConversionWithoutBoundsCheckConfig extends TaintTracking::Confi
+   */
+  int getBitSize() { this = TMkMaxValueState(result, _) }
+
+  /**
+   * Gets whether the architecture is 32 bit or 64 bit, if that has been
+   * determined already.
+   */
+  int getArchitectureBitSize() {
+    this = TMkMaxValueState(_, TMk32Bit()) and result = 32
+    or
+    this = TMkMaxValueState(_, TMk64Bit()) and result = 64
+  }
+
+  /**
+   * Gets the bitsize we should use for a sink.
+   *
+   * If the architecture bit size is known, then we should use that. Otherwise,
+   * we should use 32 bits, because that will lead to more results.
+   */
+  bindingset[default]
+  int getSinkBitSize(int default) {
+    if this = TMkMaxValueState(_, TMk64Bit()) then result = 64 else result = default
+  }
+
+  /** Gets a textual representation of this element. */
+  string toString() {
+    exists(string suffix |
+      if exists(this.getArchitectureBitSize())
+      then suffix = " (on " + this.getArchitectureBitSize() + "-bit architecture)"
+      else suffix = ""
+    |
+      result = "MaxValueState(max value <= 2^(" + this.getBitSize() + ")-1" + suffix
+    )
+  }
+}
+
+/**
+ * A node that blocks some flow states and transforms some others as they flow
+ * through it.
+ */
+abstract class BarrierFlowStateTransformer extends DataFlow::Node {
+  /**
+   * Holds if this should be a barrier for `flowstate`.
+   *
+   * This includes flow states which are transformed into other flow states.
+   */
+  abstract predicate barrierFor(MaxValueState flowstate);
+
+  /**
+   * Gets the flow state that `flowstate` is transformed into.
+   *
+   * Due to limitations of the implementation the transformation defined by this
+   * predicate must be idempotent, that is, for any input `x` it must be that:
+   * ```
+   * transform(transform(x)) = transform(x)
+   * ```
+   */
+  abstract MaxValueState transform(MaxValueState flowstate);
+}
+
+/**
+ * A node that is safely guarded by an `UpperBoundCheckGuard`.
+ *
+ * When this guarantees that a variable in the non-constant operand is less
+ * than some value this may be a barrier guard which should block some flow
+ * states and transform some others as they flow through.
+ */
+class UpperBoundCheck extends BarrierFlowStateTransformer {
+  UpperBoundCheckGuard g;
+
+  UpperBoundCheck() {
+    this = DataFlow::BarrierGuard<upperBoundCheckGuard/3>::getABarrierNodeForGuard(g)
   }
 
-/** Gets the bit size of the source. */
-int getSourceBitSize(IntegerConversionFlowState state) { state = TFlowstate(_, result, _) }
+  override predicate barrierFor(MaxValueState flowstate) {
+    g.isBoundFor2(flowstate.getBitSize(), flowstate.getSinkBitSize(32))
+  }
+
+  override MaxValueState transform(MaxValueState state) {
+    this.barrierFor(state) and
+    result.getBitSize() =
+      max(int bitsize |
+        bitsize = validBitSize() and
+        bitsize < state.getBitSize() and
+        not g.isBoundFor2(bitsize, state.getSinkBitSize(32))
+      |
+        bitsize
```

Comment:
```suggestion
```

When there is only one aggregation variable of the form `Type name` then the default expression is `name` and can be dropped. See https://codeql.github.com/docs/ql-language-reference/expressions/#omitting-parts-of-an-aggregation